File: python/greedy/gas-station.py
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        worststart = 0
        minimumnet = gas[0] - cost[0]
        totalnet = 0
        for i in range(len(gas)):
            totalnet += gas[i]
            totalnet -= cost[i]
            if totalnet <= minimumnet:
                worststart = i
                minimumnet = totalnet

        if totalnet < 0:
            return -1

        if worststart == len(gas) - 1:
            return 0
        else:
            return worststart + 1

File: c++/greedy/gas-station.cc
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        std::vector<int> net(gas.size(), std::numeric_limits<int>::max());
        int current_net = 0;
        int start = 0;
        for (int i = 0; i < gas.size(); ++i) {
            current_net += gas[i] - cost[i];
            net[i] = current_net;
            if (current_net < net[start]) {
                start = i;
            }
        }
        
        if (current_net < 0) {
            return -1;
        }

        return (start == gas.size() - 1 ? 0 : start + 1);
    }
};
